### PR TITLE
Iter 5

### PR DESCRIPTION
**Pull request by llravell**
Сервис сокращения url. В рамках ПРа реализовано:
- конфигурация сервиса через аргументы командной строки и переменные окружения
- базовые роуты для сохранения урла и редиректа
- тесты роутера

### CODE REVIEW

File: internal/api/handlers.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package api
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+)
+
+type URLStorage interface {
```
Comment (llravell) [inline internal/api/handlers.go]: Стоит ли интерфейсы зависимостей размещать в том же файле? Есть какие-то хорошие практики на этот счет?

Comment (arttet) [reply internal/api/handlers.go]: https://github.com/evrone/go-clean-template/blob/master/internal/usecase/interfaces.go

File: cmd/shortener/main.go [hunk 2]
```diff
@@ -1,3 +1,42 @@
...
+	p := config.MakeParseStrategy(
+		config.NewArgsParser(c),
+		config.NewEnvParser(c),
+	)
+
+	err := p.Parse()
+	return c, err
+}
+
+func runServer(c *config.Config) error {
```
Comment (arttet) [inline cmd/shortener/main.go]: [SUGGESTION] If I were you, I would move it to the `app` package.
https://github.com/golang-standards/project-layout/tree/master/internal/app/_your_app_

File: internal/api/handlers.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	Save(hash string, url string)
+	Get(hash string) (string, error)
+}
+
+type HashGenerator interface {
+	Generate(len int) string
+}
+
+func makeSaveURLHandler(s URLStorage, hg HashGenerator, baseAddr string) http.HandlerFunc {
```
Comment (arttet) [inline internal/api/handlers.go]: [CHANGE REQUEST] It should be a method.

File: internal/api/handlers.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+		hash := hg.Generate(10)
+		s.Save(hash, string(url))
+
+		w.WriteHeader(http.StatusCreated)
+		w.Write([]byte(fmt.Sprintf("%s/%s", baseAddr, hash)))
+	}
+}
+
+func makeResolveURLHandler(s URLStorage) http.HandlerFunc {
```
Comment (arttet) [inline internal/api/handlers.go]: [CHANGE REQUEST] It should be a method.

File: internal/api/router.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package api
+
+import (
+	"github.com/go-chi/chi/v5"
+)
+
+func BuildRouter(s URLStorage, hg HashGenerator, baseAddr string) chi.Router {
```
Comment (arttet) [inline internal/api/router.go]:
```suggestion
func Router(s URLStorage, hg HashGenerator, baseAddr string) chi.Router {
```

File: internal/config/parsers.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	return env.Parse(p.c)
+}
+
+type unionParser struct {
+	parsers []Parser
+}
+
+var _ Parser = (*unionParser)(nil)
+
+func (up *unionParser) Parse() error {
```
Comment (arttet) [inline internal/config/parsers.go]: [CLARIFICATION] Why do you need it?

File: internal/api/handlers.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+func makeSaveURLHandler(s URLStorage, hg HashGenerator, baseAddr string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		res, err := io.ReadAll(r.Body)
+		url := string(res)
+		if err != nil || url == "" {
+			http.Error(w, "Bad request", http.StatusBadRequest)
+			return
+		}
+
+		hash := hg.Generate(10)
```
Comment (arttet) [inline internal/api/handlers.go]: [CHANGE REQUEST] This is a magic number.

**Review by arttet — CHANGES_REQUESTED** [review]
- [x] Split the application into layers
https://github.com/golang-standards/project-layout
https://medium.com/@deanrubin/the-three-layered-architecture-fe30cb0e4a6#:~:text=The%20presentation%20layer%20is%20the,Update%20or%20Delete)%20operation%20flow.
- [x] Please integrate CodeCov https://about.codecov.io/blog/getting-started-with-code-coverage-for-golang/
- [x] Please read and use the Conventional Commits specification https://www.conventionalcommits.org/en/v1.0.0/
- [x] Please integrate the linter https://github.com/golangci/golangci-lint-action
- [x] Please delete deprecated constraints in CI
https://github.com/llravell/go-shortener/blob/bd734cd734d03b0eabad2342bdc49eb5e74810a7/.github/workflows/shortenertest.yml#L65

File: .github/workflows/shortenertest.yml [hunk 8]
```diff
@@ -61,124 +61,6 @@ jobs:
...
-          github.head_ref == 'iter16' ||
-          github.head_ref == 'iter17' ||
-          github.head_ref == 'iter18' ||
-          github.head_ref == 'iter19' ||
-          github.head_ref == 'iter20' ||
-          github.head_ref == 'iter21' ||
-          github.head_ref == 'iter22' ||
-          github.head_ref == 'iter23' ||
-          github.head_ref == 'iter24'
-        run: |
```
Comment (arttet) [inline .github/workflows/shortenertest.yml]: [CHANGE REQUEST] Please revert these changes. You have to run these tests.

File: .github/workflows/test-coverage.yml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 2
+
+      - uses: actions/setup-go@v2
+        with:
+          go-version: '1.22'
```
Comment (arttet) [inline .github/workflows/test-coverage.yml]:
```suggestion
          go-version: stable
```

File: .github/workflows/test-coverage.yml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  pull_request:
+  push:
+    branches:
+      - main
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (arttet) [inline .github/workflows/test-coverage.yml]:
```suggestion
      - uses: actions/checkout@v4
```

File: .github/workflows/test-coverage.yml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+          go-version: '1.22'
+
+      - name: Install dependencies
+        run: go mod download
+
+      - name: Run coverage
+        run: go test -race -coverprofile=coverage.out -covermode=atomic ./...
+
+      - name: Upload coverage to Codecov
+        uses: codecov/codecov-action@v3
```
Comment (arttet) [inline .github/workflows/test-coverage.yml]:
```suggestion
        uses: codecov/codecov-action@v4
```

**Review by arttet — CHANGES_REQUESTED** [review]
- [x] Please recover autotests

File: internal/controller/httpv1/url_test.go [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/llravell/go-shortener/internal/usecase"
```
Comment (arttet) [inline internal/controller/httpv1/url_test.go:13]: [CHANGE REQUEST] Please sort imports
https://stackoverflow.com/questions/38704769/golang-import-grouping-by-package

**Review by arttet — APPROVED** [review]

File: internal/controller/httpv1/url_test.go (hunk 12 above)
Comment (llravell) [reply internal/controller/httpv1/url_test.go:13]: Они вроде корректно отсортированы, поставил goimports, он не стал менять тут порядок

**Pull request merged**